OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs

```diff
@@ -0,0 +1,328 @@
+				tempParticle.SwingDirection *= -1;
+
+			// Perform the X-Axis-Swing
+			tempParticle.SwingOffset += tempParticle.SwingDirection * tempParticle.SwingSpeed;
+			return tempParticle;
+		}
+
+		public void Tick(Actor self)
+		{
+			windTickCountdown--;
+		}
+
+		Particle WindLogic(Particle tempParticle)
+		{
+			if (!info.ChangingWind)
+				targetWindXOffset = info.WindStrength[0];
+			else if (windTickCountdown <= 0)
+			{
+				windTickCountdown = Game.CosmeticRandom.Next(info.WindTick[0], info.WindTick[1]);
+				if (Game.CosmeticRandom.Next(2) == 1 && currentWindIndex > 0)
+				{
+					currentWindIndex--;
+					targetWindXOffset = info.WindStrength[currentWindIndex];
+				}
+				else if (currentWindIndex < info.WindStrength.Length - 1)
+				{
+					currentWindIndex++;
+					targetWindXOffset = info.WindStrength[currentWindIndex];
+				}
+			}
+
+			// Fading the wind in little steps towards the TargetWindOffset
+			if (info.InstantWindChanges)
+				currentWindXOffset = targetWindXOffset;
+			else if (currentWindXOffset != targetWindXOffset)
+			{
+				if (currentWindXOffset > targetWindXOffset)
+					currentWindXOffset -= 0.00001f;
+				else if (currentWindXOffset < targetWindXOffset)
+					currentWindXOffset += 0.00001f;
+			}
+
+			return tempParticle;
+		}
+
+		Particle Movement(Particle tempParticle)
+		{
+			tempParticle.PosX += tempParticle.DirectionScatterX + tempParticle.SwingOffset + currentWindXOffset;
+			tempParticle.PosY += tempParticle.Gravity;
+			return tempParticle;
+		}
+
+		// AntiScroll keeps the particels in place when scrolling the viewport
+		Particle AntiScroll(Particle tempParticle, WorldRenderer wr)
+		{
+			if (antiScrollPrevTopLeft != wr.Viewport.TopLeft)
+			{
+				if (antiScrollPrevTopLeft.X != wr.Viewport.TopLeft.X)
+					tempParticle.PosX += antiScrollPrevTopLeft.X - wr.Viewport.TopLeft.X;
+
+				if (antiScrollPrevTopLeft.Y != wr.Viewport.TopLeft.Y)
+					tempParticle.PosY += antiScrollPrevTopLeft.Y - wr.Viewport.TopLeft.Y;
+			}
+
+			return tempParticle;
+		}
+
+		Particle EdgeCheckReplace(Particle tempParticle, WorldRenderer wr)
+		{
+			// X-Axis
+			tempParticle.PosX %= Game.Renderer.Resolution.Width;
+			if (tempParticle.PosX < 0)
+				tempParticle.PosX += Game.Renderer.Resolution.Width;
+
+			// Y-Axis
+			tempParticle.PosY %= Game.Renderer.Resolution.Height;
+			if (tempParticle.PosY < 0 && particleCountFader != ParticleCountFaderType.FadeIn)
+				tempParticle.PosY += Game.Renderer.Resolution.Height;
+
+			return tempParticle;
+		}
+
+		void UpdateWeatherOverlay(WorldRenderer wr)
+		{
+			ParticlesCountLogic(wr);
+
+			for (var i = 0; i < ParticleList.Count; i++)
+			{
+				Particle tempParticle = ParticleList[i];
+
+				tempParticle = XAxisSwing(tempParticle);
+				tempParticle = WindLogic(tempParticle);
+				tempParticle = Movement(tempParticle);
+				tempParticle = AntiScroll(tempParticle, wr);
+				tempParticle = EdgeCheckReplace(tempParticle, wr);
```

Comment:
Consider making all the methods return void and pass the particle by reference. This avoids having to copy the (rather large) struct. I found this change made ```UpdateWeatherOverlay``` twice as fast.

There seem to be some comments earlier regarding the mutability of the struct, so I'm assuming it was decided we needed a mutable struct and not a class.